cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,384 @@
+
+        if self.qubit_map != base_op.qubit_map:
+            base_repr += f'.with_qubit_mapping({dict_repr(self.qubit_map)})'
+        if self.measurement_key_map != base_op.measurement_key_map:
+            base_repr += (
+                '.with_measurement_key_mapping(' + f'{dict_repr(self.measurement_key_map)})'
+            )
+        if self.param_resolver != base_op.param_resolver:
+            base_repr += f'.with_params({proper_repr(self.param_resolver)})'
+        if self.repetitions != 1:
+            base_repr += f'.repeat({self.repetitions})'
+        return base_repr
+
+    def __str__(self):
+        # TODO: support out-of-line subcircuit definition in string format.
+        header = self.circuit.serialization_key()
+        msg_lines = str(self.circuit).split('\n')
+        msg_width = max([len(header) - 4] + [len(line) for line in msg_lines])
+        circuit_msg = '\n'.join(
+            ['[ {line:<{width}} ]'.format(line=line, width=msg_width) for line in msg_lines]
+        )
+        args = []
+        base_op = self.base_operation()
+
+        def dict_str(d: Dict) -> str:
+            pairs = [f'{k}: {v}' for k, v in sorted(d.items())]
+            return '{' + ', '.join(pairs) + '}'
+
+        if self.qubit_map != base_op.qubit_map:
+            args.append(f'qubit_map={dict_str(self.qubit_map)}')
+        if self.measurement_key_map != base_op.measurement_key_map:
+            args.append(f'key_map={dict_str(self.measurement_key_map)}')
+        if self.param_resolver != base_op.param_resolver:
+            args.append(f'params={self.param_resolver.param_dict}')
+        if self.repetitions != 1:
+            args.append(f'loops={self.repetitions}')
+        if not args:
+            return f'{header}\n{circuit_msg}'
+        return f'{header}\n{circuit_msg}({", ".join(args)})'
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': 'CircuitOperation',
+            'circuit': self.circuit,
+            'repetitions': self.repetitions,
+            # JSON requires mappings to have keys of basic types.
+            # Pairs must be sorted to ensure consistent serialization.
+            'qubit_map': sorted(self.qubit_map.items()),
+            'measurement_key_map': self.measurement_key_map,
+            'param_resolver': self.param_resolver,
+        }
+
+    @classmethod
+    def _from_json_dict_(
+        cls, circuit, repetitions, qubit_map, measurement_key_map, param_resolver, **kwargs
+    ):
+        result = (
+            cls(circuit)
+            .with_qubit_mapping(dict(qubit_map))
+            .with_measurement_key_mapping(measurement_key_map)
+            .with_params(param_resolver)
+        )
+        if repetitions != 1:
+            return result.repeat(repetitions)
+        return result
+
+    # Methods for constructing a similar object with one field modified.
+
+    def with_circuit(self, new_circuit: 'cirq.AbstractCircuit') -> 'CircuitOperation':
```

Comment:
This was intended as an analogue to `GateOperation.with_gate`, but on further inspection that method is used exclusively in other `GateOperation` methods for which the analogy breaks down - e.g. `_resolve_parameters_` in `GateOperation` uses `with_gate`, but since `CircuitOperation` defers parameter resolution it doesn't need to call `with_circuit`.

I'd be happy to remove this if you don't see any other use case for it.